loki/transformations/utilities.py

```diff
@@ -623,20 +639,94 @@ def find_driver_loops(section, targets):
         List of subroutines that are to be considered as part of
         the transformation call tree.
     """
+    class FindDriverLoops(Visitor):
+        """
+        A  visitor that collects all driver loops in a section of the IR.
+
+        If there are nested loops, then the highest level pragma-marked driver
+        loop will take precedence and be considered the driver loop. If there
+        is no pragma-marked driver loop, then the highest level loop that does
+        not contain a driver loop will be considered the driver loop.
+        """
+
+        def __init__(self):
+            super().__init__()
+            self.driver_loops = []
 
-    driver_loops = []
-    nested_driver_loops = []
-    for loop in FindNodes(ir.Loop).visit(section):
-        if loop in nested_driver_loops:
-            continue
+        def visit_tuple(self, o, **kwargs):
+            nested_pragma_loop = False
+            for i in o:
+                nested_pragma_loop |= self.visit(i, **kwargs)
+            return nested_pragma_loop
```

Comment:
Not a big fan of the mixing of return values via pass-by-reference lists and actual return value. I prefer returning a tuple of the multiple return values, with explicit combination of the values where required, such as here for the tuple:

```suggestion
        @classmethod
        def default_retval(cls):
            return False, [], []

        def visit_tuple(self, o, **kwargs):
            nested_pragma_loop = False
            nested_target_loops = []
            nested_target_calls = []
            for i in o:
                retval = self.visit(i, **kwargs)
                nested_pragma_loop |= retval[0]
                nested_target_loops += retval[1]
                nested_target_calls += retval[2]
            return nested_pragma_loop, nested_target_loops, nested_target_calls
```